File: LeetCode/77.Combinations/Mario/77.Combinations.java
class Solution {

    private List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combine(int n, int k) {
        backtracking(new ArrayList<>(), k, 1, n);
        return res;
    }

    private void backtracking(List<Integer> list, int k, int range, int n) {
        if (k == 0) {
            res.add(new ArrayList<>(list));
            return;
        }

        for (int i = range; i <= n; i++) {
            list.add(i);
            backtracking(list, k - 1, i + 1, n);
            list.remove(list.size() - 1);
        }
    }
}

File: LeetCode/77.Combinations/Mario/77.Combinations.py
#!/usr/bin/env python3


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:

        res = []

        def backtracking(i, path):
            if i == k:
                res.append(path[:])
                return

            # select num from i+1 to n 1 -> 4 / 2 -> 4
            for j in range(i + 1, n + 1):
                path.append(j)
                # print(path)
                backtracking(j + 1, path)
                path.pop()

        backtracking(0, [])

        return res
